File: medium/next-permutation.py
from typing import List

class Solution:
	def nextPermutation(self, nums: List[int]) -> None:
		i = len(nums) - 2

		# Find where the increasing sequence is broken, starting from the right
		while i >= 0: 
			if nums[i] < nums[i + 1]:
				break
			i -= 1

		if i >= 0: # This means we found a pivot (i.e: where the increasing sequence is broken starting from the right)
			# If we've found a pivot, then find the smallest number that is greater than the number at the pivot, in the whole array
			x = len(nums) - 1

			while nums[x] <= nums[i] and x >=0:
				x -= 1

			# So then the smallest number that is greater than the number at the pivot is at x
			# Since we're navigating from right to left, when we reach the break point, the entries right to it are already sorted in descending order
			nums[x], nums[i] = nums[i], nums[x]

		self.reverseFrom(nums, i + 1)

		print(nums)
	
	def reverseFrom(self, nums: List[int], idx: int) -> None:
		idy = len(nums) - 1

		while idx <= idy:
			nums[idx], nums[idy] = nums[idy], nums[idx]
			idx += 1
			idy -= 1

solution = Solution()
sample = [5,1,1]

ans = solution.nextPermutation(sample)

File: medium/next-permutation.ts
// https://leetcode.com/problems/next-permutation/
/**
 Do not return anything, modify nums in-place instead.
 */
function nextPermutation(nums: number[]): void {
    // Initialize the pivot/breakpoint
    var pivot = nums.length - 2;

    // Find the current pivot/breakpoint
    // i.e where nums[pivot + 1] > nums[pivot] 
    while(nums[pivot] >= nums[pivot+1]){
        pivot -= 1;
    }
    
    // Now that we've found our pivot;
    //  - Find the index of next number in the list that is > than the number at the pivot
    
    if(pivot >= 0){
        var i = nums.length - 1
        while(nums[pivot] >= nums[i]){
            i -= 1;
        }
    

        // Once we find the index of the next number that is > than the number at the pivot
        // - We swap them
        const x = nums[pivot]
        const y = nums[i]

        nums[pivot] = y
        nums[i] = x
    }

    // After swapping:
    // - Reverse the array from the next position after the pivot

    reverseNums(nums, pivot + 1)  
};

const reverseNums = (nums: number[], idx: number) => {
    var right = nums.length - 1

    while(idx <= right) {
        const x = nums[idx]
        const y = nums[right]

        nums[idx] = y
        nums[right] = x

        right -= 1;
        idx += 1
    }
}
